docs/rules/extensions.md

```diff
@@ -2,7 +2,7 @@
 
 <!-- end auto-generated rule header -->
 
-Some file resolve algorithms allow you to omit the file extension within the import source path. For example the `node` resolver can resolve `./foo/bar` to the absolute path `/User/someone/foo/bar.js` because the `.js` extension is resolved automatically by default. Depending on the resolver you can configure more extensions to get resolved automatically.
+Some file resolve algorithms allow you to omit the file extension within the import source path. For example the `node` CommonJS resolver can resolve `./foo/bar` to the absolute path `/User/someone/foo/bar.js` because the `.js` extension is resolved automatically by default. Depending on the resolver you can configure more extensions to get resolved automatically.
```

Comment:
it's a node resolver, not a CJS resolver - it's just that it doesn't yet support "exports" or ESM.

Reply:
To confirm we're on the same page here. My understanding is that `node` in this context refers specifically to NodeJS, is that correct?

NodeJS documents two separate module loaders, one for CommonJS and one for ECMAScript. The ESM loader explicitly says:

> _It does no extension searching. A file extension must be provided when the specifier is a relative or absolute file URL._

https://nodejs.org/api/packages.html#modules-loaders

Reply:
yes, but "node" describes many node versions over a long period of time, it's not inherently "node latest". It's a node resolver because it matches node's resolution *at one point* and because it intends to match the latest version eventually.

Yes, the ESM loader does no extension searching, but any time it's resolving something in a package's "exports" field, the specifier need not (and often should not!) have an extension. It's only on relative files that users are forced to suffer the pain of specifying extensions.

Reply:
So this proposed change is just updating the `eslint-plugin-import` documentation to be correct now that NodeJS has more that one resolver, right? NodeJS has had more than one resolver since 2017 and referring to them as the same for this scenario is misleading.

>Yes, the ESM loader does no extension searching, but any time it's resolving something in a package's "exports" field, the specifier need not (and often should not!) have an extension. It's only on relative files that users are forced to suffer the pain of specifying extensions.

Given the rule:

```json
{
  "import/extensions": ["error", "always"],
}
```

This code is correct:

```js
import assert from "assert";

import {aFunc, bFunc} from "../local.js";
```

So `import/extensions` already correctly only enforces extensions on relative files and leaves packages untouched.

Reply:
That's not accurate - node's ESM resolver became unflagged in 2020, not before, and it wasn't until April 2021 that the last non-ESM LTS line went EOL (i was in the node modules working group; i'm quite familiar with the timeline)

Reply:
Then you'll no doubt also know that NodeJS's ESM resolver was available behind a flag and/or using `.mjs` since [v8 back in 2017](https://nodejs.org/docs/latest-v8.x/api/esm.html). That said, I'm not interested in debating timelines.

I found the `import/extensions` documentation in it's current state to be confusing as it has examples using the `import` syntax and specifically discusses "`node`" as not requiring a file extension for local files which as far as I can tell has never been the case. Because I wasted time trying this rule out and having to revert those changes - and had seen another user already raised this before (#2043) - I thought it might be useful for other users to clarify this.

Feel free to suggest alternative wording, I am happy to rephrase my changes. But I can't tell what you're looking for from your feedback so far.

Reply:
Behind a flag doesn't count :-)

Let me give it a fresh review and see what I can suggest.